src/cpp/core/Database.cpp

```diff
@@ -0,0 +1,638 @@
+      return DatabaseError(query.prepareError_.get());
+
+   try
+   {
+      query.statement_.define_and_bind();
+      bool result = query.statement_.execute(true);
+
+      if (pDataReturned)
+         *pDataReturned = result;
+
+      query.statement_.bind_clean_up();
+
+      return Success();
+   }
+   catch (soci::soci_error& error)
+   {
+      return DatabaseError(error);
+   }
+}
+
+Error Connection::execute(Query& query,
+                          Rowset& rowset)
+{
+   if (query.prepareError_)
+      return DatabaseError(query.prepareError_.get());
+
+   try
+   {
+      query.statement_.define_and_bind();
+      query.statement_.exchange_for_rowset(soci::into(rowset.row_));
+      query.statement_.execute(false);
+
+      rowset.query_ = query;
+
+      return Success();
+   }
+   catch (soci::soci_error& error)
+   {
+      return DatabaseError(error);
+   }
+}
+
+Error Connection::executeStr(const std::string& queryStr)
+{
+   try
+   {
+      // SOCI backends do not necessarily support running multiple statements
+      // in one invocation - to work around this, we split any passed in SQL
+      // into one invocation per SQL statement (delimited by ;)
+      std::vector<std::string> queries;
+      boost::regex regex(";[ \\t\\r\\f\\v]*\\n");
+      std::string queryStrCopy = queryStr;
+      boost::regex_split(std::back_inserter(queries), queryStrCopy, regex);
+      for (std::string& query : queries)
+      {
+         query = string_utils::trimWhitespace(query);
+         if (!query.empty())
+            session_ << query;
+      }
+
+      return Success();
+   }
+   catch (soci::soci_error& error)
+   {
+      return DatabaseError(error);
+   }
+}
+
+std::string Connection::driverName() const
+{
+   return session_.get_backend_name();
+}
+
+PooledConnection::PooledConnection(const boost::shared_ptr<ConnectionPool>& pool,
+                                   const boost::shared_ptr<Connection>& connection) :
+   pool_(pool),
+   connection_(connection)
+{
+}
+
+PooledConnection::~PooledConnection()
+{
+   pool_->returnConnection(connection_);
+}
+
+Query PooledConnection::query(const std::string& sqlStatement)
+{
+   return connection_->query(sqlStatement);
+}
+
+Error PooledConnection::execute(Query& query,
+                              Rowset& rowset)
+{
+   return connection_->execute(query, rowset);
+}
+
+Error PooledConnection::execute(Query& query,
+                                bool* pDataReturned)
+{
+   return connection_->execute(query, pDataReturned);
+}
+
+Error PooledConnection::executeStr(const std::string& queryStr)
+{
+   return connection_->executeStr(queryStr);
+}
+
+std::string PooledConnection::driverName() const
+{
+   return connection_->driverName();
+}
+
+boost::shared_ptr<IConnection> ConnectionPool::getConnection()
+{
+   // block until a connection is available
+   boost::shared_ptr<Connection> connection;
+   connections_.deque(&connection, boost::posix_time::pos_infin);
```

Comment:
This will hang until there is a connection available in the pool, which is what I intended. And like you said, they are handed out via shared ptr to help ensure that they are not leaked.

Reply:
Right, I know that hanging until one is available is intended. The concern would be around hanging forever if for some reason the connections are all never returned. Do you think it would be worth setting some very long timeout and then logging an error and exiting? That way it will be easier to find such an issue, in the unlikely event that one every occurs?

Reply:
I think exiting is a little heavy handed, but I think logging that there's a potential hang isn't a bad idea. I'll add this.